File: 81. Search in Rotated Sorted Array II.cpp
/*
Follow up for "Search in Rotated Sorted Array":
What if duplicates are allowed?

Would this affect the run-time complexity? How and why?
Suppose an array sorted in ascending order is rotated at some pivot unknown to you beforehand.

(i.e., 0 1 2 4 5 6 7 might become 4 5 6 7 0 1 2).

Write a function to determine if a given target is in the array.

The array may contain duplicates.

Subscribe to see which companies asked this question
*/
class Solution {
public:
	bool search(vector<int>& nums, int target) {
		if (nums.size() != 0)
		{
			unsigned int first = 0;
			unsigned int last = nums.size() - 1;
			do
			{
				unsigned int mid = (first + last) / 2;
				if (nums[mid] == target)
					return true;

				if (nums[first] < nums[mid])
					if (nums[first] <= target && target <= nums[mid])
						last = mid - 1;
					else
						first = mid + 1;
				else if (nums[first] == nums[mid])
					first++;
				else
					if (nums[mid] < target && target <= nums[last])
						first = mid + 1;
					else
						last = mid - 1;
			} while (first != last + 1);
			return false;
		}
		else
			return false;
	}
};

File: Sorting and Searching/81. Search in Rotated Sorted Array II.py
# Suppose an array sorted in ascending order is rotated at some pivot unknown to you beforehand.
#
# (i.e., [0,0,1,2,2,5,6] might become [2,5,6,0,0,1,2]).
#
# You are given a target value to search. If found in the array return true, otherwise return false.
#
# Example 1:
#
# Input: nums = [2,5,6,0,0,1,2], target = 0
# Output: true
# Example 2:
#
# Input: nums = [2,5,6,0,0,1,2], target = 3
# Output: false
# Follow up:
#
# This is a follow up problem to Search in Rotated Sorted Array, where nums may contain duplicates.
# Would this affect the run-time complexity? How and why?

class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: bool
        """
        return self.helper(nums, target, 0, len(nums) - 1)

    def helper(self, nums, target, left, right):
        # [2, 2, 2, 3, 4, 2]
        if left > right:
            return False
        mid = left + (right - left) // 2
        if nums[mid] == target:
            return True

        if nums[left] < nums[mid]:  # left side is normal
            # if target is in the left side, search on the left
            if nums[left] <= target < nums[mid]:
                return self.helper(nums, target, left, mid - 1)
            # else search on the right side
            else:
                return self.helper(nums, target, mid + 1, right)
        elif nums[mid] < nums[right]:  # right side is normal
            # if target is in the right side, search on the right side
            if nums[mid] < target <= nums[right]:
                return self.helper(nums, target, mid + 1, right)
            # else search on the left side
            else:
                return self.helper(nums, target, left, mid - 1)
        else:  # both side are not normal
            # first search on the left side
            res = self.helper(nums, target, left, mid - 1)
            # not find, search on the right side
            if not res:
                return self.helper(nums, target, mid + 1, right)
            # return ans
            else:
                return res
